sycl/doc/design/CompileTimeProperties.md

```diff
@@ -596,6 +582,56 @@ we do not allow non-fundamental types, how do we represent properties like
 allow `std::tuple`, where the type of each element is one of the fundamental
 types listed above.
 
+### IR representation via `@llvm.global.annotations`
+
+Properties that are implemented using
+`[[__sycl_detail__::add_ir_attributes_global_variable()]]`, are represented in LLVM IR
+as elements in the global variable `@llvm.global.annotations`. `@llvm.global.annotations` is an array of `{ ptr, ptr, ptr, i32, ptr}` structs, where each struct represents an annotation on a global pointer. The values of each field of the struct are as follows:
+
+* pointer to global object to be annotated, global variable in our case. If there are nested annotations on the global variable, constant `getelementptr` instructions can be used to specify at which level the annotation is applied.
+* pointer to global string, representing the annotation. Each global string is a list of `{name:value}` pairs from the `[[__sycl_detail__::add_ir_attributes_global_variable()]]` annotation. Property values are converted to strings
+in the same way as described above, except that the `nullptr` value is represented as an empty string, `{name}`.
+* pointer to global string, representing the path of the source file where the annotation was applied.
+* i32 integer representing the line number in the source file where the annotation was applied. 
+* ptr to annotation arguments, `null` pointer in our case 
+
+```cpp
+device_global<fpga_mem<int[4], decltype(properties(word_size<2>))>, decltype(properties(host_access_none))> dg;
+```
+
+Below is what the LLVM IR will look like for the SYCL code above.
+
+```llvm
+%"fpga_mem" = type { [4 x i32] }
+%"device_global" = type { %"fpga_mem" }
+
+@.str = private unnamed_addr addrspace(1) constant [25 x i8] c"{5826:DEFAULT}{5884:2}\00"
+@.str.1 = private unnamed_addr addrspace(1) constant [14 x i8] c"example.cpp\00"
+@.str.2 = private unnamed_addr addrspace(1) constant [11 x i8] c"{6168:0}\00"
+
+@dg = dso_local local_unnamed_addr addrspace(1) global { %"device_global" } zeroinitializer
+@llvm.global.annotations = appending global 
```

Comment:
I had a concern that passes like `GlobalDCE` could remove the variable, but it seems like `appending` linkage will do the trick to preserve it (even though it is not explicitly mentioned anywhere, had to go look into source code).

However, there is still a question of an extra "waste" produced by our compiler. What would happen to this global variable during SPIR-V translation? Will it be removed or translated as-is?

Reply:
`@llvm.global.annotations` is a special "intrinsic" global variable, that SPIRV is already aware of. It does not appear as a global variable, but each element in the array gets translated into a decoration.